### PR TITLE
feat: rewrite pipe

### PR DESCRIPTION
**Pull request by PhotonQuantum**

### CODE REVIEW

File: src/opts.rs [hunk 1]
```diff
@@ -150,4 +160,6 @@ pub struct Opts {
     pub concurrent_resolve: usize,
     #[structopt(flatten)]
     pub transfer_config: TransferConfig,
+    #[structopt(flatten)]
+    pub rewrite_config: RewriteConfig,
```
Comment (skyzh) [inline src/opts.rs:164]: Is it necessary to set rewrite config in cli? From my perspective, only some repos need hard-coded rewrite configurations.

File: src/rewrite_pipe.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        if byte_stream.length > self.max_length {
+            Ok(byte_stream)
+        } else {
+            match byte_stream.object {
+                ByteObject::LocalFile {
+                    ref mut file,
+                    ref path,
+                } => {
+                    if let (Some(file), Some(path)) = (file, path) {
+                        let mut buffer = String::new();
```
Comment (skyzh) [inline src/rewrite_pipe.rs:72]: Can we ensure what stored in file is UTF-8 encoded?

File: src/rewrite_pipe.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                        file.read_to_string(&mut buffer).await?;
+                        let content = (*self.rewrite_fn)(buffer)?.into_bytes();
+                        let content_length = content.len() as u64;
+
+                        let original_filename = path
+                            .file_name()
+                            .ok_or_else(|| {
+                                Error::ProcessError(String::from("given path is not a file"))
```
Comment (skyzh) [inline src/rewrite_pipe.rs:80]: Use `StorageError` or create something like `PipeError`

File: src/rewrite_pipe.rs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                            })?
+                            .to_str()
+                            .ok_or_else(|| {
+                                Error::ProcessError(String::from("corrupted filename"))
+                            })?;
+                        let path = format!(
```
Comment (skyzh) [inline src/rewrite_pipe.rs:86]: I think it's easier to just append `.rewrite` to filename here. We don't really care about what's inside the cache path.

File: src/rewrite_pipe.rs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                        let mut f = BufWriter::new(
+                            OpenOptions::default()
+                                .create(true)
+                                .truncate(true)
+                                .write(true)
+                                .read(true)
+                                .open(&path)
+                                .await?,
+                        );
+                        f.write_all(content.as_ref()).await?;
```
Comment (skyzh) [inline src/rewrite_pipe.rs:101]: `&content`

File: src/rewrite_pipe.rs (hunk 4 above)
Comment (skyzh) [reply src/rewrite_pipe.rs:86]: Or we could just modify the original file?

File: src/utils.rs [hunk 6]
```diff
@@ -83,6 +85,17 @@ pub fn rewrite_snapshot(target_snapshot: &mut [SnapshotPath]) {
     }
 }
 
+pub fn fn_regex_rewrite(
+    pattern: String,
+    rewrite: String,
+) -> Box<dyn Fn(String) -> Result<String> + Sync + Send> {
+    Box::new(move |data| {
+        let re = regex::Regex::new(pattern.as_str())
```
Comment (skyzh) [inline src/utils.rs:93]: Create regex every time we call this function may lead to a huge overhead. Consider create this regex upon loading config.

**Review by skyzh — COMMENTED** [review]
LGTM overall

File: src/opts.rs (hunk 1 above)
Comment (PhotonQuantum) [reply src/opts.rs:176]: Users may want to use their own rewriting rules to adapt their own mirror service.

For example, SJTUG may safely specify something like `--rewrite-pattern https://download.haskell.org/ --rewrite-target https://mirrors.sjtu.edu.cn/haskell/`, but users may want to specify their own domains or rewriting rules.

Comment (skyzh) [reply src/opts.rs:176]: From my perspective, this kind of config should be part of the subcommand. e.g. `./mirror-clone ghc-up --target= https://mirrors.sjtu.edu.cn/haskell/`. And we developers should decide on what pattern to rewrite.

File: src/rewrite_pipe.rs (hunk 2 above)
Comment (PhotonQuantum) [reply src/rewrite_pipe.rs:84]: I didn't get it. If the file isn't encoded as UTF-8, `read_to_string` will raise an io::Error. Are you suggesting adding support for other encodings?

Comment (skyzh) [reply src/rewrite_pipe.rs:84]: If so, there will be an error when we get this object, and this file will not be transferred.

File: src/rewrite_pipe.rs [hunk 7]
```diff
@@ -78,55 +80,37 @@ where
             match byte_stream.object {
                 ByteObject::LocalFile {
                     ref mut file,
-                    ref path,
+                    path: _,
```
Comment (skyzh) [inline src/rewrite_pipe.rs:83]: You can use `..` here.

File: src/rewrite_pipe.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                    ref mut file,
+                    path: _,
+                } => {
+                    if let Some(ref mut file) = file {
+                        let mut buffer = String::new();
+                        if file.read_to_string(&mut buffer).await.is_err() {
+                            warn!(logger, "rewrite_pipe: not a valid UTF-8 file, ignored");
+                            Ok(byte_stream)
+                        } else {
+                            match (*self.rewrite_fn)(buffer) {
```
Comment (skyzh) [inline src/rewrite_pipe.rs:91]: Could we use static dispatch for `rewrite_fn`?

File: src/utils.rs [hunk 9]
```diff
@@ -76,11 +79,15 @@ pub fn rewrite_url_string(url_encode_map: &[(&'static str, &'static str)], key:
     key
 }
 
-pub fn rewrite_snapshot(target_snapshot: &mut [SnapshotPath]) {
-    let gen_map = generate_s3_url_encode_map();
-    for path in target_snapshot {
-        path.0 = rewrite_url_string(&gen_map, &path.0);
-    }
+pub fn fn_regex_rewrite(
+    pattern: Regex,
```
Comment (skyzh) [inline src/utils.rs:83]: Should be `&Regex`.

**Review by skyzh — APPROVED** [review]
LGTM overall. You may merge at your wish.

**Pull request merged**